In [1]:
# if the email is spam , spam = 1 ; else 0 . 

In [2]:
import math
import numpy as np 
import pandas as pd 
"""import nltk 
from nltk.corpus import stopwords"""
import string
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
from sklearn.feature_extraction.text import CountVectorizer , TfidfVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import Pipeline

In [3]:
df = pd.read_csv("emails.csv")
print(df.head())

                                                text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1


In [4]:
#print the shape
df.shape

(5728, 2)

In [5]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(df["text"],df["spam"],test_size = 0.20,random_state=0)


In [6]:
# I converted it to np array, which was giving a dictionary key error. 

x_train_np = np.array(X_train)
y_train_np = np.array(y_train)

x_test_np = np.array(X_test)
y_test_np = np.array(y_test)

len(y_train_np)

4582

In [7]:
#                                       SUMMARIZE ALL FORMULAS

#                         To find conditional probablity formula for unique words . 



#                              (        counting of "X" words in spam/ham + 1           )
#     P(X| ham or spam) =     --------------------------------------------------------------
#                             (When row is spam/ham , count of all words + Unique word count)



#       P(spam | Test(i)) = P(X1| spam) * P(X2|  spam) * ..... * P(X2|  spam) * P_Spam

#       P(ham | Test(i)) = P(X1| ham) * P(X1| ham) * ..... * P(X1| ham) * P_Ham

#       Whichever is larger, we make the classification.

In [8]:
# To find , List has only Ham rows . 

ham_only = list()
ham_indices = list()
for i in range(len(x_train_np)):
    if y_train_np[i] == 0:
        ham_only.append(x_train_np[i])
        ham_indices.append(i)
        
# To find , List has only Spam rows .

spam_only = list()
spam_indices = list()
for i in range(len(x_train_np)):
    if y_train_np[i] == 1:
        spam_only.append(x_train_np[i])
        spam_indices.append(i)

        
        
# ham_only : a list of only ham rows.
# spam_only : a list of only spam rows.
# ham/spam_indices : a list of indices of only ham/spam rows. To find , the total frequency of a word in a type (where?)



In [9]:
# When row is ham , count of all words 

def ham_all_words_count(onlyham):
    
    # I appended here all splitted words.
    list_all_ham_words = list()
    
    for i in onlyham:
        
        list_all_ham_words += i.split(" ")
        
        
    # I appended here all word again without punctuation. Because i used np.array and np.array gum up my data with punc.
    new_list_all_ham_words = list()
    
    for i in list_all_ham_words:
        if i not in string.punctuation:
            new_list_all_ham_words.append(i)
    
    # Conclusion i returned here , count of all words when mail is ham. 
    return len(new_list_all_ham_words)
    
    
ham_all_words_count(ham_only)



895519

In [10]:
# When row is spam , count of all words

def spam_all_words_count(onlyspam):
    
    # I appended here all splitted words.
    list_all_spam_words = list()
    
    for i in onlyspam:

        list_all_spam_words += i.split(" ")
        

    # I appended here all word again without punctuation. Because i used np.array and np.array gum up my data with punc.
    new_list_all_spam_words = list()

    for i in list_all_spam_words:
        if i not in string.punctuation:
            new_list_all_spam_words.append(i)
    

    # Conclusion i returned here , count of all words when mail is ham. 
    return len(new_list_all_spam_words)


spam_all_words_count(spam_only)

227250

In [11]:
# buraya mat.plot ile iki türün kelime sayılarını grafikleştirebilirsin

In [12]:
# We counted here " How many HAM type of mail are there here ? "

sum_of_ham=0
for i in range(len(y_train_np)):
    if y_train_np[i] == 0 :
        sum_of_ham+=1
print("sum_of_ham : " , sum_of_ham)




# We counted here " How many SPAM type of mail are there here ? "

sum_of_spam=0
for i in range(len(y_train_np)):
    if y_train_np[i] == 1 :
        sum_of_spam+=1
print("sum_of_spam : " , sum_of_spam)



# For the formula : we compute here (sum of ham OR spam) / all mail counts
p_ham = sum_of_ham / len(y_train_np)
print("P_Ham : " , p_ham)
p_spam = sum_of_spam / len(y_train_np)
print("P_Spam : " , p_spam)

sum_of_ham :  3475
sum_of_spam :  1107
P_Ham :  0.7584024443474465
P_Spam :  0.24159755565255348


In [13]:
# ------------------------------------------- PART 2 -----------------------------------------------------

# i used sklearn.CountVectorizer library
# it is returned some words to create BoW. 

vectorizer = CountVectorizer(min_df=0.01)
   
X = vectorizer.fit_transform(x_train_np)

vectorizer2 = CountVectorizer(ngram_range=(2, 2),min_df=0.01)
X2 = vectorizer2.fit_transform(x_train_np)
unique_words2 = vectorizer2.get_feature_names()

# unique words name's list        ( BAG OF WORDS )
unique_words = vectorizer.get_feature_names()
print("                              ---     BAG OF WORDS (UNIQUE WORDS LIST)     ---         ")
print()
print(unique_words)
print()

# unique words frequency list for each row
each_row_word_frequency = X.toarray()
each_row_word_frequency2 = X2.toarray()


# --------------------------TF-IDF-----------------------------------

# here for part3 implementing tf idf only ham words

ham_tf = Pipeline([('count', CountVectorizer(min_df=0.01)),
                ('tfid', TfidfTransformer())]).fit(ham_only)
ham_tf['count'].transform(ham_only).toarray()

ham_tf_olasilik = ham_tf['tfid'].idf_
print(ham_tf_olasilik)

ham_tf_unique = ham_tf["count"].get_feature_names()



#  to find the most useful and useless words when guessing

ham_tf_yararli = sorted(ham_tf_olasilik,reverse=True)[9]

ham_tf_yararsiz = sorted(ham_tf_olasilik)[9]

#--------------------------------------------------------------------

# here for part3 implementing tf idf only spam words

spam_tf = Pipeline([('count', CountVectorizer(min_df=0.01)),
                ('tfid', TfidfTransformer())]).fit(spam_only)
spam_tf['count'].transform(spam_only).toarray()

spam_tf_olasilik = spam_tf['tfid'].idf_
print(spam_tf_olasilik)

spam_tf_unique = spam_tf["count"].get_feature_names()



#  to find the most useful and useless words when guessing

spam_tf_yararli = sorted(spam_tf_olasilik,reverse=True)[9]

spam_tf_yararsiz = sorted(spam_tf_olasilik)[9]

# --------------------------TF-IDF-----------------------------------



print("How many unique words are there?  : ", len(unique_words))

# i did this because we have to calculate words frequency seperately...
hamlarin_label_arrayi = list()
hamlarin_label_arrayi2 = list()

for i in ham_indices:
    hamlarin_label_arrayi.append(each_row_word_frequency[i])
    hamlarin_label_arrayi2.append(each_row_word_frequency2[i])

spamlarin_label_arrayi = list()
spamlarin_label_arrayi2 = list()

for j in spam_indices:
    spamlarin_label_arrayi.append(each_row_word_frequency[j])
    spamlarin_label_arrayi2.append(each_row_word_frequency2[j])


                              ---     BAG OF WORDS (UNIQUE WORDS LIST)     ---         

['00', '000', '01', '02', '03', '04', '05', '0500', '06', '0600', '07', '08', '09', '10', '100', '1092', '11', '12', '13', '14', '1400', '15', '150', '16', '17', '18', '19', '1998', '1999', '20', '200', '2000', '2001', '2002', '2004', '2005', '21', '212', '215', '22', '23', '24', '25', '250', '254', '26', '27', '28', '29', '30', '300', '31', '32', '33', '34', '345', '348', '35', '36', '37', '38', '3848', '39', '40', '41', '42', '43', '44', '4473', '45', '46', '47', '471', '48', '49', '50', '500', '51', '512', '52', '5290', '53', '54', '55', '56', '57', '58', '59', '60', '62', '646', '650', '66', '70', '710', '713', '75', '76798', '77002', '80', '800', '853', '90', '95', '98004', '99', 'ability', 'able', 'about', 'above', 'absolutely', 'academic', 'accept', 'accepted', 'access', 'according', 'account', 'accounting', 'accounts', 'accuracy', 'achieve', 'across', 'act', 'action', 'activities', 'actual'

[2.7369052  4.30160722 5.49007584 ... 5.26181719 5.26181719 3.86537046]
[3.55596457 2.96688675 4.57632466 ... 1.35873999 4.54457596 3.8359246 ]
How many unique words are there?  :  1926


In [14]:
# I calculated here , the frequency of a word in HAM / SPAM mail only.

def sum_of_frequency_each_uniq(frekans_icin_array,ham_spam_only):
    sirali_frekans = list()
    for i in range(len(unique_words)):
        frekans = 0
        for j in range(len(ham_spam_only)):
            frekans += frekans_icin_array[j][i]
        sirali_frekans.append(frekans)
    return sirali_frekans


ham_frequency_list = sum_of_frequency_each_uniq(hamlarin_label_arrayi,ham_only)

print("                   HAM FREQUENCY")
print()
print()
print()
print(ham_frequency_list)

print()
print()
print()
print()
print()
print()
print()
print()
print()
print()
print()
print()

print("                   SPAM FREQUENCY")
print()
print()
print()
spam_frequency_list = sum_of_frequency_each_uniq(spamlarin_label_arrayi,spam_only)

print(spam_frequency_list)


                   HAM FREQUENCY



[1310, 372, 1392, 1023, 1018, 1205, 846, 69, 678, 82, 669, 942, 848, 1758, 135, 75, 1329, 1186, 373, 481, 107, 691, 33, 553, 505, 472, 563, 63, 146, 718, 57, 3860, 2527, 56, 8, 5, 388, 111, 143, 394, 455, 407, 442, 18, 155, 342, 459, 419, 455, 1187, 40, 337, 155, 162, 115, 59, 125, 155, 132, 133, 113, 81, 127, 239, 113, 105, 117, 305, 75, 235, 146, 143, 101, 96, 129, 236, 83, 114, 163, 142, 153, 101, 108, 141, 114, 111, 131, 141, 69, 26, 83, 154, 27, 42, 154, 848, 58, 78, 81, 54, 71, 353, 52, 46, 78, 132, 117, 433, 1575, 240, 21, 79, 62, 55, 458, 73, 136, 101, 56, 25, 27, 114, 79, 102, 102, 90, 86, 110, 89, 161, 296, 447, 53, 51, 109, 83, 111, 54, 46, 3, 29, 44, 103, 45, 479, 223, 469, 88, 180, 9, 123, 107, 86, 280, 112, 79, 34, 57, 202, 1602, 153, 60, 70, 35, 150, 298, 1275, 54, 116, 116, 3834, 231, 149, 111, 96, 111, 2371, 375, 229, 64, 16734, 0, 81, 139, 277, 56, 46, 75, 103, 294, 137, 1782, 124, 159, 60, 210, 68, 46, 56, 200, 126, 85, 45, 304, 86

[281, 505, 54, 28, 30, 8, 23, 4, 18, 1, 31, 17, 17, 336, 180, 0, 69, 110, 28, 50, 0, 87, 49, 43, 33, 52, 168, 10, 3, 144, 35, 23, 21, 44, 133, 309, 75, 0, 0, 28, 12, 92, 120, 37, 3, 19, 44, 20, 48, 153, 25, 31, 23, 23, 25, 4, 0, 32, 41, 15, 8, 0, 36, 39, 6, 39, 18, 15, 0, 56, 80, 7, 0, 22, 43, 202, 102, 15, 2, 11, 0, 5, 16, 16, 8, 89, 23, 42, 49, 36, 1, 1, 49, 88, 2, 0, 85, 0, 0, 91, 31, 1, 124, 209, 0, 99, 44, 61, 347, 88, 41, 5, 25, 15, 136, 27, 317, 13, 34, 28, 24, 16, 96, 30, 15, 37, 12, 37, 14, 18, 82, 294, 2, 116, 13, 5, 5, 22, 52, 84, 74, 77, 0, 35, 150, 3, 85, 35, 0, 60, 13, 13, 13, 40, 8, 0, 54, 57, 3, 803, 40, 24, 29, 25, 26, 64, 215, 33, 15, 49, 170, 30, 45, 0, 45, 80, 623, 24, 8, 22, 5165, 63, 3, 0, 0, 9, 29, 10, 21, 55, 20, 474, 63, 36, 0, 18, 16, 8, 7, 66, 48, 5, 49, 10, 3, 9, 5, 15, 64, 15, 5, 1266, 49, 14, 38, 0, 2, 5, 8, 6, 978, 0, 33, 15, 9, 7, 40, 48, 48, 6, 13, 9, 24, 14, 5, 17, 805, 10, 0, 2, 7, 0, 29, 15, 1, 40, 30, 188, 50, 19, 23, 80, 101, 4, 16, 132, 28, 33, 23

                             *********************    PART 1   *******************

I can give simple 3 example words...

"2000"    3860 times in ham   |  5 times in spam
"2001"    2527 times in ham   |  23 times in spam
"zimin"   403 times in ham    |  0 times in spam

Yes, it is quite convenient to apply this algorithm to data processing. I will easily calculate what type of text it is by calculating conditional probabilities.

I've already thrown the highest frequency words that are only in spam and raw into a list. from here it is also very simple that "2000" named word's frequency is 3860 in ham (but in spam only 5 times),"2001"  named word's frequency is 2527 in ham(but in spam only 23 times) and "zimin" named word's frequency is 403 in ham (but in spam only 0 times) seem to be quite effective when making predictions. 

As a result, such decisive words can be easily seen even by looking at the list.

------------------------ Stop Words---------------------------------
It's not a good thing that a word passes too often on both sides . There are english words called stop_words. (a , an, to , the , one ,and ,with... as such ) we can say that the most ineffective words for the three of these words that occur most often are.

                             *********************    PART 1   *******************

In [15]:
# I calculated here , bayes of each word. ( seperately for both ham and spam )

def cond_probablity(param1,param2):
    conditional_list = list()
    for i in range(len(unique_words)):
        cond_prob = (param2[i] + 1) / (param1 + len(unique_words))
        conditional_list.append(cond_prob)

    return conditional_list

#                              the formula

#        (        counting of "x" words in spam/ham + 1           )
#        --------------------------------------------------------------
#        (When row is spam/ham , count of all words + Unique word count)



print("PROBABLITY FOR HAM WORDS : ")
print()
ham_con_prob_list = cond_probablity(ham_all_words_count(ham_only),ham_frequency_list)
print(ham_con_prob_list[:50])               # i printed only 50 element because so that you can see more comfortably


print()
print()
print()
print()

print("PROBABLITY FOR SPAM WORDS : ")
print()
spam_con_prob_list = cond_probablity(spam_all_words_count(spam_only),spam_frequency_list)
print(spam_con_prob_list[:50])              # i printed only 50 element here so that you can see more comfortably


PROBABLITY FOR HAM WORDS : 

[0.0014608137546033462, 0.00041562435581010535, 0.0015521842564168277, 0.0011410169982561606, 0.001135445626194363, 0.0013438149413055954, 0.0009437904272685234, 7.799920886516722e-05, 0.0007565923259921221, 9.248477622584114e-05, 0.0007465638562808863, 0.0010507607708550384, 0.0009460189760932424, 0.0019600086913404164, 0.00015154132008089632, 8.468485533932441e-05, 0.0014819849684381773, 0.0013226437274707643, 0.00041673863022246487, 0.0005370802667572943, 0.00012034163653482944, 0.000771077893352796, 3.788533002022408e-05, 0.0006173080244471806, 0.0005638228526539231, 0.0005270517970460586, 0.0006284507685707759, 7.131356239101004e-05, 0.00016379833861685116, 0.0008011633024865033, 6.462791591685285e-05, 0.004302213506120152, 0.0028168857144448963, 6.351364150449331e-05, 1.0028469711235786e-05, 6.6856464741571905e-06, 0.00043345274640785785, 0.00012479873418426755, 0.00016045551537977257, 0.00044013839288201503, 0.0005081091320359465, 0.00045462396024268

In [16]:
# I transferred the words as KEY and the possibilities as VALUE to the DICTIONARY because it was difficult to access.

def list_to_dict(param1, param2):

    return dict(list(zip(param1, param2)))

ham_sozluk = list_to_dict(unique_words, ham_con_prob_list)

spam_sozluk = list_to_dict(unique_words, spam_con_prob_list)

ham_sozluk_tfidf = list_to_dict(ham_tf_unique, ham_tf_olasilik)

spam_sozluk_tfidf = list_to_dict(spam_tf_unique, spam_tf_olasilik)

print("Just some for example  : " ,ham_sozluk["his"])

Just some for example  :  0.0010273610081954882


In [17]:
#                                                MODEL 
#
#                         Here I am sending test data to the model i have built


# I have broken down the test data into words for the test. (row by row)

def test_row_split():
    sentences_word_list = list()
    
    for i in x_test_np:
        sentences_word_list.append(i.split(" "))
    
    return sentences_word_list
        

splitted_test_mails = test_row_split()




# To implement Bayes, I added the conditional probabilities logarithmically that I calculated earlier.
# I've found the last possibility. So I can calculate the accuracy by comparing it with its real value (spam or raw?).

def bayes(p_x , x_sozluk):
    all_bayes = list()
    for mail in splitted_test_mails:
        each_bayes = math.log(p_x)
        for word in mail[1:]:
            if word not in x_sozluk:
                each_bayes*=1
            else:
                each_bayes += math.log(x_sozluk['{}'.format(word)])
        all_bayes.append(each_bayes)
    return all_bayes
           
    
   #  with count vectorizer 
    
ham_bayes = bayes(p_ham,ham_sozluk)
spam_bayes = bayes(p_spam,spam_sozluk)

    
   #  with tf-idf  
   
ham_bayes_tfidf = bayes(p_ham, ham_sozluk_tfidf)
spam_bayes_tfidf = bayes(p_spam, spam_sozluk_tfidf)


    
    
    # I counted here true predictions . And i sent a list these . (for library of metrics )   

def predict(ham, spam):
    prediction_list = list()

    for i in range(len(x_test_np)):
        if ham[i] > spam[i]:
            prediction_list.append(0)
        else:
            prediction_list.append(1)

    return prediction_list


list_of_predict = predict(ham_bayes, spam_bayes)


list_of_predict_tfidf = predict(ham_bayes_tfidf,spam_bayes_tfidf)




##### CAUTION ###### 
# If you want to see prediction label and real value , you can remove comment line(3) BELOW and run it .


# print("prediction | real value")
# for j in range(len(y_test_np)):
#     print(list_of_predict[j] , "         |       " , y_test_np[j])
    


In [18]:
#--------------------------------------------- PART 4 ----------------------------------      for UNIGRAM

# Accuracy  -  Precision  -  Recall  -  F1Score 

print("PART 4 \n")
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# from sklearn.metrics import recall_score
def metrics(y_testt, predict_lists):
    accuracyy = accuracy_score(y_testt, predict_lists)

    print("Accuracy Score :    ½", accuracyy * 100)

    precision = precision_score(y_testt, predict_lists)

    print("Precision Score :   %", precision * 100)

    recall = recall_score(y_testt, predict_lists)

    print("Recall Score :   %", recall * 100)

    f1 = f1_score(y_testt, predict_lists)

    print("F1 Score :   %", f1 * 100)
    
print("Count Vectorizer")
print()
metrics(y_test_np, list_of_predict)
print()
print()
print("TF-IDF")
print()
metrics(y_test_np,list_of_predict_tfidf)


PART 4 

Count Vectorizer

Accuracy Score :    ½ 97.99301919720767
Precision Score :   % 94.73684210526315
Recall Score :   % 96.55172413793103
F1 Score :   % 95.63567362428843


TF-IDF

Accuracy Score :    ½ 91.36125654450262
Precision Score :   % 76.47058823529412
Recall Score :   % 89.65517241379311
F1 Score :   % 82.53968253968254


In [19]:
# --------------------------------------   PART 3  ---------------------------------------------
# 

#   I have listed the words with the highest probability when finding these words. (with sorted()).
#   Then I kept the indexes of these values in a list and called the words that are the owners of these possibilities.
#   Here we can easily see that meaningless numbers have reduced the state of guessing to "Ham" .
#   In the same way, the presence of some words was effective in determining the class as SPAM / HAM

print("PART 3 \n")

yararli_ham = list()
for i in ham_tf_unique:
    if ham_sozluk_tfidf.get(i) >= ham_tf_yararli:
        yararli_ham.append(i)
        
print("List the 10 words whose presence most strongly predicts that the mail is ham.")
print("Useful Words For Ham : \n", yararli_ham[:10])
print()
print()

yararsiz_ham = list()
for i in ham_tf_unique:
    if ham_sozluk_tfidf.get(i) >= ham_tf_yararsiz:
        yararsiz_ham.append(i)
        
print("List the 10 words whose absence most strongly predicts that the mail is ham.")
print("Useless Words For Ham : \n",yararsiz_ham[:10])
print()
print()



yararli_spam = list()
for i in spam_tf_unique:
    if spam_sozluk_tfidf.get(i) >= spam_tf_yararli:
        yararli_spam.append(i)
        
print("List the 10 words whose presence most strongly predicts that the mail is spam.")
print("Useful Words For Spam : \n",yararli_spam[:10])
print()
print()

yararsiz_spam = list()
for i in spam_tf_unique:
    if spam_sozluk_tfidf.get(i) >= spam_tf_yararsiz:
        yararsiz_spam.append(i)
        
print("List the 10 words whose absence most strongly predicts that the mail is spam.")
print("Useless Words For Spam : \n ",yararsiz_spam[:10])
print()
print()
print()

print("I showed here English Stop Words \n")
print(ENGLISH_STOP_WORDS)

#   The most commonly used words in English that I mention at the beginning are stop_words, such as conjunctions, numbers.
# I have listed these words below . If we remove these words from "bag_of_words",
# our accuracy value will increase significantly. Because the MOST BASIC LOGIC of this classification is that there 
# is a lot of difference in word frequencies. In other words, if the frequency of the same word in HAM 
# and SPAM is close to each other, it will not be of much use to us at the prediction stage.

PART 3 

List the 10 words whose presence most strongly predicts that the mail is ham.
Useful Words For Ham : 
 ['100038', 'approve', 'bambos', 'behind', 'copyright', 'courses', 'cunningham', 'degree', 'described', 'difficulties']


List the 10 words whose absence most strongly predicts that the mail is ham.
Useless Words For Ham : 
 ['00', '000', '0011', '01', '02', '03', '04', '05', '0500', '06']


List the 10 words whose presence most strongly predicts that the mail is spam.
Useful Words For Spam : 
 ['1995', '218', '55', '58', '69', 'accepted', 'across', 'actually', 'advises', 'amonq']


List the 10 words whose absence most strongly predicts that the mail is spam.
Useless Words For Spam : 
  ['00', '000', '0000', '01', '02', '03', '05', '06', '07', '10']



I showed here English Stop Words 

frozenset({'most', 'becoming', 'yet', 'anywhere', 'herein', 'often', 'afterwards', 'here', 'hasnt', 'only', 'too', 'eight', 'part', 'else', 'per', 'mill', 'found', 'amount', 'then', 'cant', 'pl

In [20]:
# -------------------------- BIGRAM -------------

print("BAG OF WORDS FOR BIGRAM")
print(unique_words2)

BAG OF WORDS FOR BIGRAM
['00 am', '00 pm', '01 2000', '01 2001', '01 pm', '02 2000', '02 2001', '04 2001', '05 01', '05 2000', '06 2000', '07 2000', '08 2000', '09 2000', '10 00', '10 000', '10 2000', '10 30', '10 am', '10 pm', '1092 fax', '11 00', '11 2000', '11 2001', '11 30', '11 am', '12 00', '12 2000', '12 pm', '14 2000', '14 pm', '1400 smith', '15 2000', '16 2000', '16 2001', '17 2000', '18 2000', '19 th', '20 2000', '2000 01', '2000 02', '2000 03', '2000 04', '2000 05', '2000 06', '2000 07', '2000 08', '2000 09', '2000 10', '2000 11', '2000 12', '2001 01', '2001 02', '2001 03', '2001 04', '2001 05', '2001 06', '2001 07', '2001 08', '2001 09', '2001 10', '2001 11', '2001 12', '21 2000', '22 2000', '23 2000', '23 2001', '24 2000', '25 2000', '254 710', '26 2000', '27 2000', '28 2000', '29 2000', '30 2000', '30 2001', '30 am', '30 pm', '31 2000', '31 pm', '39 pm', '44 20', '4473 office', '45 pm', '47 am', '512 471', '52 am', '710 1092', '710 4473', '713 345', '713 348', '713 646', 